L69-cpp-core-event-manager.md

```diff
@@ -0,0 +1,282 @@
+EventManager API
+----
+* Author(s): @muxi
+* Approver: @markdroth
+* Status: Draft
+* Implemented in: gRPC Core
+* Last updated: June 1, 2020
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+This proposal describes a plan to upgrade gRPC's event engine to a newer set of interfaces called EventManager API.The gRPC team is making a fundamental change in the gRPC core's polling model. Instead of borrowing threads from the application and giving the application control over which threads poll which fds, gRPC will migrate to a new model based on EventManager, which provides dedicated polling threads with an internal thread pool. User callbacks and timers will also be handled by the EventManager. The EventManager interface will be a public API, which allows users to integrate gRPC into the event loops of their applications if necessary.
+
+## Background
+The gRPC has been using a custom event handling framework called iomgr API. The original iomgr API had only two implementations, one for posix (Linux and MacOS) and one for Windows. As part of an effort to improve performance, we changed the posix iomgr implementation to support pluggable polling engines, selectable at run-time.  We then wrote [a whole bunch of different polling engines](https://github.com/grpc/grpc/blob/master/doc/core/grpc-polling-engines.md) to experiment on and determine which one(s) worked best.  Some of them didn't work out and have been deleted, but we still support three of them (see below).
+
+The EventManager interface aims at simplifying the usage and maintenance of gRPC event framework. It allows users of gRPC to write their own event management implementation to be used by gRPC for specific needs. The project also paves the way for supporting features such as the C++ callback API in gRPC (See proposal [gRFC L67: C++ callback-based asynchronous API](https://github.com/grpc/proposal/pull/180)).
+
+## Proposal
+The following event manager interfaces will be available in the gRPC core.
+
+```cpp
+// An external handle to an EventManager task, which can be used to cancel the 
+// task. The EventManager task is added via RunFunction.
+struct TaskKey;
```

Comment:
I'm a little unclear on how `TaskKey` works. It's an empty struct, but it's also an output parameter of `RunFunction`. Does that mean that `RunFunction` has recorded the address of the `TaskKey` that I passed in? If that's the case, then should `TaskKey` be non-copyable and non-movable? (Perhaps also non-assignable).

I have the same question for `AlarmKey`.

Was there any consideration of having `RunFunction` and company return `TaskKey` instances as the handles?

Reply:
Each implementation will define their own TaskKey and AlarmKey struct as they can be different. The caller of `RunFunction` provides the struct and `RunFunction` fills TaskKey contents in. I did not really understand the "record the address" part of your question.

Reply:
Are `TaskKey` and `AlarmKey` base types for opaque handles? If so, should the parameter to `RunFunction` and `ScheduleAlarmAt` be `TaskKey**` and `AlarmKey**` respectively? Cancellation would need to be changed to take `TaskKey*` and `AlarmKey*` parameters too.

Assuming that the empty struct is intentional, what would help me is an example showing how cancellation will work. Right now, I don't see any way to cancel a function.

```lang-cpp
std::shared_ptr<BaseEventManagerInterface> em = GetEventManager();

TaskKey tk;
em->RunFunction(some_handler, PRIORITY_LOW, &tk);

// Now what?
// No member functions on TaskKey...
// I don't see a CancelTask member function...
```

Alarms looks more usable:

```lang-cpp
void SomeFun() {
    std::shared_ptr<BaseEventManagerInterface> em = GetEventManager();

    AlarmKey ak;
    em->ScheduleAlarmAt(in_the_future, some_handler, PRIORITY_LOW, &ak, some_slack);

    DoSomethingElse();

    em->ConsistentCancel(ak);
}
```

However, `AlarmKey` is an empty struct, so there's nowhere for the EM implementation to store any state inside of `AlarmKey`. The only thing the EM can use to identify the alarm is the address of the `AlarmKey` instance that was passed in to `ScheduleAlarmAt`. (Unless I'm missing something...) That _can_ work, but then the following will break in a subtle way:

```lang-cpp
void SomeFunc() {
    std::shared_ptr<BaseEventManagerInterface> em = GetEventManager();
    
    AlarmKey ak;
    em->ScheduleAlarmAt(in_the_future, some_handler, PRIORITY_LOW, &ak, some_slack);
    
    // Need to capture em and ak by value as we're returning from
    // SomeFunc() after scheduling the callback.
    // AlarmKey has a copy ctor, so I assume it's OK to copy it.
    ScheduleCallbackForLaterUsingSomeOtherCallbackSystem([ak, em] {

        // 💥 AlarmKey doesn't have value semantics... Making a copy wasn't OK...
        // The event manager was actually using the address of AlarmKey and this ak
        // has a different address.
        em->ConsistentCancel(ak);
    });
}
```

If the intention is to use the address of the TaskKey/AlaramKey as the identity for the event manager, then can copy, move, and assignment be disabled?

```lang-cpp
struct AlarmKey {
    AlarmKey(const AlarmKey &) = delete;
    AlarmKey& operator=(const AlarmKey &) = delete;

    // IIRC, these aren't needed but some people add them to signal that
    // non-movable was intentional.
    AlarmKey(AlarmKey &&) = delete;
    AlarmKey& operator=(AlarmKey &&) = delete;
};
```

That will force the caller to keep the address stable:

```lang-cpp
void SomeFunc() {
    std::shared_ptr<BaseEventManagerInterface> em = GetEventManager();
    
    auto ak = std::make_unique<AlarmKey>();
    em->ScheduleAlarmAt(in_the_future, some_handler, PRIORITY_LOW, ak.get(), some_slack);
    
    // This uses C++14 capture. For C++11, could .release() into a pointer, capture that,
    // then create a unique_ptr inside the body of the lambda. Or use a shared_ptr.
    ScheduleCallbackForLaterUsingSomeOtherCallbackSystem([ak = std::move(ak), em] {
        em->ConsistentCancel(*ak);
    });
}
```